io/EptAddonWriter.cpp

```diff
@@ -0,0 +1,252 @@
+*       with the distribution.
+*     * Neither the name of Hobu, Inc. or Flaxen Geo Consulting nor the
+*       names of its contributors may be used to endorse or promote
+*       products derived from this software without specific prior
+*       written permission.
+*
+* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+* "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+* LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+* FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+* COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+* INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+* BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+* OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+* AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+* OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
+* OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY
+* OF SUCH DAMAGE.
+****************************************************************************/
+
+#include "EptAddonWriter.hpp"
+
+#include <cassert>
+
+#include <json/json.h>
+
+#include <arbiter/arbiter.hpp>
+
+#include "private/EptSupport.hpp"
+
+namespace pdal
+{
+
+namespace
+{
+    const StaticPluginInfo s_info
+    {
+        "writers.ept_addon",
+        "EPT Writer",
+        "http://pdal.io/stages/writers.ept.html",
+        { "ept_addon", "ept-addon" }
+    };
+
+    std::string getTypeString(Dimension::Type t)
+    {
+        const auto base(Dimension::base(t));
+        if (base == Dimension::BaseType::Signed) return "signed";
+        if (base == Dimension::BaseType::Unsigned) return "unsigned";
+        if (base == Dimension::BaseType::Floating) return "float";
+        throw std::runtime_error("Invalid dimension type");
+    }
+}
+
+CREATE_STATIC_STAGE(EptAddonWriter, s_info)
+
+EptAddonWriter::EptAddonWriter()
+    : m_addonsArg(new Json::Value())
+{ }
+
+EptAddonWriter::~EptAddonWriter()
+{ }
+
+std::string EptAddonWriter::getName() const { return s_info.name; }
+
+void EptAddonWriter::addArgs(ProgramArgs& args)
+{
+    args.add("addons", "Mapping of output locations to their dimension names",
+            *m_addonsArg);
+    args.add("threads", "Number of worker threads", m_numThreads);
+}
+
+void EptAddonWriter::addDimensions(PointLayoutPtr layout)
+{
+    m_nodeIdDim = layout->registerOrAssignDim("EptNodeId",
+            Dimension::Type::Unsigned32);
+    m_pointIdDim = layout->registerOrAssignDim("EptPointId",
+            Dimension::Type::Unsigned32);
+}
+
+void EptAddonWriter::ready(PointTableRef table)
+{
+    m_arbiter.reset(new arbiter::Arbiter());
+    m_pool.reset(new Pool(std::max<uint64_t>(m_numThreads, 4)));
+
+    MetadataNode meta(table.privateMetadata("ept"));
+
+    if (meta.findChild("info").value().empty())
+        throwError("EPT writer must descend from EPT reader");
```

Comment:
> Doesn't seem any different from having no info or keys, which doesn't get checked.

This is checking the `info` though - and since `info` and `keys` are added at the same time by the reader, the presence of `info` is used to indicate the presence of the EPT reader.

> Do you even need to report.

If there's no `info` and thus no `EPT reader` which is required for this writer, what behavior would be preferable to this `throw`?

Reply:
Fixed error message in 111d9a0.